File: loom-tools/src/loom_tools/tokens/failure_counts.py
"""Per-account consecutive-failure counter for token rotation.

Tracks consecutive ``TOKEN_EXHAUSTED`` failures per account so the
spawn wrapper can auto-unpin a stuck pin. The counter is stored at
``.loom/tokens/.failure_counts`` as JSON::

    {"<account_name>": {"count": <int>, "last_failure": "<ISO8601 UTC>"}}

The counter is reset on:

* a successful spawn for that account (``record_success``),
* any operator allowlist mutation (``pin``, ``unpin``, ``add``, ``remove``).

Reaching the threshold (default 5) is the signal that the operator's
pin must be auto-released by the wrapper. The threshold check is
``>= 5`` — a 6th failure does not bump the count past the trigger; it
still triggers (idempotent at-or-above).

Concurrency:
    Writes are guarded by a sibling ``mkdir``-lock and use the
    "atomic rewrite" pattern (``mktemp`` + ``os.replace``). Concurrent
    bash and Python writers (e.g. spawn-claude.sh and the operator
    CLI) are coordinated via the same lock dir.

Public API:
    record_failure(workspace, name, *, threshold=5) -> int
    record_success(workspace, name) -> None
    reset_all(workspace) -> None
    get_count(workspace, name) -> int
    threshold_reached(workspace, name, *, threshold=5) -> bool
    DEFAULT_THRESHOLD
"""

from __future__ import annotations

import json
import os
import tempfile
from datetime import datetime, timezone
from pathlib import Path

# Reuse the lock primitive from bad_tokens (mkdir-only, macOS-safe).
from loom_tools.tokens.bad_tokens import _MkdirLock

DEFAULT_THRESHOLD = 5


def _tokens_dir(workspace: Path | str) -> Path:
    return Path(workspace) / ".loom" / "tokens"


def _state_path(workspace: Path | str) -> Path:
    return _tokens_dir(workspace) / ".failure_counts"


def _lock_path(workspace: Path | str) -> Path:
    return _tokens_dir(workspace) / ".failure_counts.lock"


def _ensure_dir(workspace: Path | str) -> Path:
    d = _tokens_dir(workspace)
    d.mkdir(parents=True, exist_ok=True)
    return d


def _read_state(path: Path) -> dict[str, dict[str, object]]:
    """Read the JSON state file, returning ``{}`` on any failure.

    A malformed or unreadable file is treated as an empty state — better
    to lose history than to refuse to record new failures.
    """
    try:
        raw = path.read_text(encoding="utf-8")
    except (FileNotFoundError, OSError):
        return {}
    if not raw.strip():
        return {}
    try:
        loaded = json.loads(raw)
    except json.JSONDecodeError:
        return {}
    if not isinstance(loaded, dict):
        return {}
    # Filter out malformed entries defensively
    cleaned: dict[str, dict[str, object]] = {}
    for k, v in loaded.items():
        if not isinstance(k, str) or not isinstance(v, dict):
            continue
        count = v.get("count", 0)
        if not isinstance(count, int) or count < 0:
            continue
        last = v.get("last_failure", "")
        if not isinstance(last, str):
            last = ""
        cleaned[k] = {"count": count, "last_failure": last}
    return cleaned


def _write_state_atomic(path: Path, state: dict[str, dict[str, object]]) -> None:
    """Write state via mktemp + os.replace (atomic on POSIX)."""
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_str = tempfile.mkstemp(
        prefix=path.name + ".",
        suffix=".tmp",
        dir=str(path.parent),
    )
    tmp = Path(tmp_str)
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            json.dump(state, fh, indent=2, sort_keys=True)
            fh.write("\n")
        os.replace(tmp, path)
    except Exception:
        try:
            tmp.unlink(missing_ok=True)
        except OSError:
            pass
        raise


def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def record_failure(
    workspace: Path | str,
    name: str,
    *,
    threshold: int = DEFAULT_THRESHOLD,
) -> int:
    """Increment the consecutive-failure counter for ``name``.

    Args:
        workspace: Repo root containing ``.loom/tokens/``.
        name: Account name (token file stem).
        threshold: Counter is capped at ``threshold`` to avoid unbounded
            growth (we only ever check ``>=``, so capping is harmless).

    Returns:
        The new count (post-increment), capped at ``threshold``.
    """
    _ensure_dir(workspace)
    path = _state_path(workspace)
    with _MkdirLock(_lock_path(workspace)):
        state = _read_state(path)
        entry = state.get(name) or {"count": 0, "last_failure": ""}
        new_count = min(int(entry.get("count", 0)) + 1, threshold)
        state[name] = {"count": new_count, "last_failure": _now_iso()}
        _write_state_atomic(path, state)
        return new_count


def record_success(workspace: Path | str, name: str) -> None:
    """Reset the counter for ``name`` (drops the key entirely).

    Args:
        workspace: Repo root containing ``.loom/tokens/``.
        name: Account name (token file stem).

    No-op if the state file is missing or the key is absent.
    """
    path = _state_path(workspace)
    if not path.is_file():
        return
    with _MkdirLock(_lock_path(workspace)):
        state = _read_state(path)
        if name not in state:
            return
        del state[name]
        if state:
            _write_state_atomic(path, state)
        else:
            try:
                path.unlink()
            except FileNotFoundError:
                pass


def reset_all(workspace: Path | str) -> None:
    """Drop all per-account counters.

    Called from the allowlist CLI after any mutation (``pin``,
    ``unpin``, ``add``, ``remove``) so a fresh allowlist starts with
    a clean slate. No-op if the state file is missing.
    """
    path = _state_path(workspace)
    if not path.is_file():
        return
    with _MkdirLock(_lock_path(workspace)):
        try:
            path.unlink()
        except FileNotFoundError:
            pass


def get_count(workspace: Path | str, name: str) -> int:
    """Return the current consecutive-failure count for ``name`` (0 if absent)."""
    path = _state_path(workspace)
    if not path.is_file():
        return 0
    state = _read_state(path)
    entry = state.get(name)
    if not entry:
        return 0
    count = entry.get("count", 0)
    return int(count) if isinstance(count, int) else 0


def threshold_reached(
    workspace: Path | str,
    name: str,
    *,
    threshold: int = DEFAULT_THRESHOLD,
) -> bool:
    """Return True iff the counter for ``name`` is ``>= threshold``."""
    return get_count(workspace, name) >= threshold


def all_counts(workspace: Path | str) -> dict[str, dict[str, object]]:
    """Return the entire state dict (for status / debugging).

    The returned dict is a snapshot — mutating it does not affect the
    persisted state.
    """
    path = _state_path(workspace)
    if not path.is_file():
        return {}
    return _read_state(path)

File: loom-daemon/src/tokens_pool/failure_counts.rs
//! Per-account consecutive-failure counter, ported from
//! `loom_tools.tokens.failure_counts`.
//!
//! Tracks consecutive `TOKEN_EXHAUSTED` failures per account so the spawn
//! wrapper can auto-unpin a stuck pin. Stored at `.loom/tokens/.failure_counts`
//! as JSON: `{"<account_name>": {"count": <int>, "last_failure": "<ISO8601 UTC>"}}`.
//!
//! The counter is reset on a successful spawn for that account
//! ([`record_success`]) or any operator allowlist mutation ([`reset_all`]).
//! Reaching the threshold (default 5) is checked `>= 5` — idempotent
//! at-or-above, matching the Python semantics exactly.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use chrono::Utc;
use serde::{Deserialize, Serialize};

use super::allowlist::atomic_write;
use super::locking::MkdirLock;
use super::paths::resolve_tokens_dir;

pub const DEFAULT_THRESHOLD: u32 = 5;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FailureEntry {
    pub count: u32,
    pub last_failure: String,
}

type State = BTreeMap<String, FailureEntry>;

fn tokens_dir(workspace: &Path) -> PathBuf {
    resolve_tokens_dir(workspace)
}

fn state_path(workspace: &Path) -> PathBuf {
    tokens_dir(workspace).join(".failure_counts")
}

fn lock_path(workspace: &Path) -> PathBuf {
    tokens_dir(workspace).join(".failure_counts.lock")
}

/// Read the JSON state file, returning an empty map on any failure — a
/// malformed/unreadable file is treated as empty rather than a hard error
/// (matches Python: better to lose history than refuse to record failures).
fn read_state(path: &Path) -> State {
    let Ok(raw) = std::fs::read_to_string(path) else {
        return State::new();
    };
    if raw.trim().is_empty() {
        return State::new();
    }
    // Parse defensively: malformed entries (non-object values, negative or
    // non-integer counts) are dropped rather than propagated.
    let Ok(serde_json::Value::Object(map)) = serde_json::from_str::<serde_json::Value>(&raw) else {
        return State::new();
    };
    let mut cleaned = State::new();
    for (k, v) in map {
        let serde_json::Value::Object(obj) = v else {
            continue;
        };
        let count = match obj.get("count") {
            Some(serde_json::Value::Number(n)) if n.is_i64() || n.is_u64() => {
                let c = n.as_i64().unwrap_or(-1);
                if c < 0 {
                    continue;
                }
                c as u32
            }
            _ => continue,
        };
        let last_failure = obj
            .get("last_failure")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        cleaned.insert(
            k,
            FailureEntry {
                count,
                last_failure,
            },
        );
    }
    cleaned
}

fn write_state(path: &Path, state: &State) -> Result<(), String> {
    let mut body = serde_json::to_string_pretty(state).map_err(|e| e.to_string())?;
    body.push('\n');
    atomic_write(path, &body)
}

fn now_iso() -> String {
    Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string()
}

/// Increment the consecutive-failure counter for `name`, capped at
/// `threshold`. Returns the new (post-increment) count.
///
/// # Errors
/// Returns an error if the lock cannot be acquired.
pub fn record_failure(workspace: &Path, name: &str, threshold: u32) -> Result<u32, String> {
    let dir = tokens_dir(workspace);
    std::fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    let path = state_path(workspace);
    let _lock = MkdirLock::acquire(&lock_path(workspace))?;
    let mut state = read_state(&path);
    let prev = state.get(name).map_or(0, |e| e.count);
    let new_count = (prev + 1).min(threshold);
    state.insert(
        name.to_string(),
        FailureEntry {
            count: new_count,
            last_failure: now_iso(),
        },
    );
    write_state(&path, &state)?;
    Ok(new_count)
}

/// Reset the counter for `name` (drops the key entirely). No-op if the state
/// file is missing or the key is absent.
///
/// # Errors
/// Returns an error if the lock cannot be acquired.
pub fn record_success(workspace: &Path, name: &str) -> Result<(), String> {
    let path = state_path(workspace);
    if !path.is_file() {
        return Ok(());
    }
    let _lock = MkdirLock::acquire(&lock_path(workspace))?;
    let mut state = read_state(&path);
    if state.remove(name).is_none() {
        return Ok(());
    }
    if state.is_empty() {
        let _ = std::fs::remove_file(&path);
    } else {
        write_state(&path, &state)?;
    }
    Ok(())
}

/// Drop all per-account counters. No-op if the state file is missing.
///
/// # Errors
/// Returns an error if the lock cannot be acquired.
pub fn reset_all(workspace: &Path) -> Result<(), String> {
    let path = state_path(workspace);
    if !path.is_file() {
        return Ok(());
    }
    let _lock = MkdirLock::acquire(&lock_path(workspace))?;
    match std::fs::remove_file(&path) {
        Ok(()) | Err(_) => Ok(()),
    }
}

/// Return the current consecutive-failure count for `name` (0 if absent).
#[must_use]
pub fn get_count(workspace: &Path, name: &str) -> u32 {
    let path = state_path(workspace);
    if !path.is_file() {
        return 0;
    }
    read_state(&path).get(name).map_or(0, |e| e.count)
}

/// Return `true` iff the counter for `name` is `>= threshold`.
#[must_use]
pub fn threshold_reached(workspace: &Path, name: &str, threshold: u32) -> bool {
    get_count(workspace, name) >= threshold
}

/// Return the entire state map (for status/debugging).
#[must_use]
pub fn all_counts(workspace: &Path) -> State {
    let path = state_path(workspace);
    if !path.is_file() {
        return State::new();
    }
    read_state(&path)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    fn make_pool() -> tempfile::TempDir {
        let tmp = tempfile::tempdir().unwrap();
        let dir = tmp.path().join(".loom").join("tokens");
        fs::create_dir_all(&dir).unwrap();
        // resolve_tokens_dir() only picks the per-repo pool when it holds at
        // least one `*.token` file — seed one so these tests deterministically
        // exercise the per-repo pool rather than falling back to this host's
        // real shared pool (~/.loom/tokens) when it is empty.
        fs::write(dir.join("seed.token"), "sk-ant-oat01-fake").unwrap();
        tmp
    }

    #[test]
    fn record_failure_increments() {
        let tmp = make_pool();
        assert_eq!(record_failure(tmp.path(), "a", 5).unwrap(), 1);
        assert_eq!(record_failure(tmp.path(), "a", 5).unwrap(), 2);
        assert_eq!(get_count(tmp.path(), "a"), 2);
    }

    #[test]
    fn record_failure_caps_at_threshold() {
        let tmp = make_pool();
        for _ in 0..10 {
            record_failure(tmp.path(), "a", 5).unwrap();
        }
        assert_eq!(get_count(tmp.path(), "a"), 5);
    }

    #[test]
    fn threshold_reached_is_at_or_above() {
        let tmp = make_pool();
        for _ in 0..5 {
            record_failure(tmp.path(), "a", 5).unwrap();
        }
        assert!(threshold_reached(tmp.path(), "a", 5));
        // A 6th failure (capped) still triggers — idempotent at-or-above.
        record_failure(tmp.path(), "a", 5).unwrap();
        assert!(threshold_reached(tmp.path(), "a", 5));
    }

    #[test]
    fn record_success_drops_key() {
        let tmp = make_pool();
        record_failure(tmp.path(), "a", 5).unwrap();
        record_success(tmp.path(), "a").unwrap();
        assert_eq!(get_count(tmp.path(), "a"), 0);
    }

    #[test]
    fn record_success_removes_file_when_last_key_dropped() {
        let tmp = make_pool();
        record_failure(tmp.path(), "a", 5).unwrap();
        record_success(tmp.path(), "a").unwrap();
        assert!(!state_path(tmp.path()).exists());
    }

    #[test]
    fn reset_all_clears_state_file() {
        let tmp = make_pool();
        record_failure(tmp.path(), "a", 5).unwrap();
        record_failure(tmp.path(), "b", 5).unwrap();
        reset_all(tmp.path()).unwrap();
        assert!(!state_path(tmp.path()).exists());
        assert_eq!(get_count(tmp.path(), "a"), 0);
    }

    #[test]
    fn malformed_state_file_reads_as_empty() {
        let tmp = make_pool();
        fs::write(state_path(tmp.path()), "not json").unwrap();
        assert_eq!(get_count(tmp.path(), "a"), 0);
        assert!(all_counts(tmp.path()).is_empty());
    }

    #[test]
    fn negative_counts_are_dropped_defensively() {
        let tmp = make_pool();
        fs::write(state_path(tmp.path()), r#"{"a": {"count": -1, "last_failure": ""}}"#).unwrap();
        assert_eq!(get_count(tmp.path(), "a"), 0);
    }
}
